### PR TITLE
Pre-IAP

### PR DESCRIPTION
**Pull request by jni2000**
Preliminary implementation of preIAP logic for early review. Still in progress, especially the dependencies on all the pre-IAP grant purging handling parts.

### CODE REVIEW

**Comment by jni2000** [comment]
we will have a new drop on this.

File: src/harness/reference_models/pre_iap/pre_iap.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+     #Creating list of cbsds from UUT FAD and Test Harness FADs
+     cbsd_list = []
+     for cbsds in uut_fad_object.getCbsds():
+        cbsd_list.append(cbsds)
+     for fad in test_harness_fad_objects:
+        for cbsds in fad.getCbsds():
+            cbsd_list.append(cbsds)
+
+     #Iterate through each CBSD and check if the FCC ID and SerialNumber matches another
+     #CBSD. If yes then check whether their grants overlap and mark the overlapping grants as duplicate.
```
Comment (piotrfilipiuk) [inline src/harness/reference_models/pre_iap/pre_iap.py]: According to the spec, we should exclude all duplicate CBSDs based on the FCC ID and SerialNumber. Meaning that at any given time a CBSD must not have active grants with more than one SAS. There is no need to check overlap among grants.

File: src/harness/reference_models/pre_iap/sas_pre_iap_fss.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    # Update sfss and pfss list
+    total_cbsd_interference = 0
+    for index, cbsd in enumerate(sfss):
+        total_cbsd_interference = total_cbsd_interference + cbsd['interference_value']
+        if total_cbsd_interference > oobe_threshold_value:
+            nc = index
+            break     
+        else:
+            nc = index + 1   
+    pfss = sfss[nc:]
```
Comment (piotrfilipiuk) [inline src/harness/reference_models/pre_iap/sas_pre_iap_fss.py]: Can you please add a comment explaining what pfss, and sfss are?

File: src/harness/reference_models/pre_iap/sas_pre_iap_fss.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+            fss_record['deploymentParam'][0]['installationParam']['latitude'],
+            fss_record['deploymentParam'][0]['installationParam']['longitude'],
+            cbsd['registrationRequest']['installationParam']['latitude'],
+            cbsd['registrationRequest']['installationParam']['longitude'])
+        if distance_km <= 40:
+            neighbouring_cbsds.append(cbsd)
+    return neighbouring_cbsds
+
+def performPurge(cbsds, fss_entities, margin_oobe):
```
Comment (piotrfilipiuk) [inline src/harness/reference_models/pre_iap/sas_pre_iap_fss.py]: I think we should a unit test for the reference implementation.

**Comment by kate-harrison** [comment]
FYI Piotr is a colleague of mine at Google.

**Comment by jni2000** [comment]
Thanks, Kate and Piotr for reviewing the PR. We will address them asap.

James Ni
Director of SAS System
www.federatedwireless.com<http://www.federatedwireless.com/>
https://www.linkedin.com/pub/james-ni/4/3b9/24a
978.996.7480
[Description: Macintosh HD:Users:Sam:Dropbox (Bluetext):Federated Wireless:Creative:Logo:Logo Lock Up:Final-Lockup:Federated Wireless No Tag:FedWireless-lockup-notag-web.png]

**Pull request closed without merging**